### PR TITLE
Re-enable attachment validation tests

### PR DESCRIPTION
**Pull request by ssoloff**
While reviewing [#2112](https://github.com/triplea-game/triplea/pull/2112/files#r128143719), I discovered that, sometime in 2012, the tests in `ValidateAttachmentsTest` were changed to no longer fail when an `IAttachment`-derived class was determined to be invalid, but rather to only report the error(s) to the console.  There are currently about 500 errors being reported, so apparently no one has noticed. :smile:  Therefore, my goal for this PR was to fix the existing errors and re-enable the tests to fail when an invalid attachment is detected.  (Of course, after reviewing this PR, the team may feel this class of tests is not even really necessary, in which case, I'm happy to convert the PR into one that removes this test fixture completely.)

The fixes mostly required adding missing methods that may be invoked via reflection (but, in reality, probably aren't; otherwise someone would have observed the failure and fixed it).  There were other fixes that required some creative changes, which I'll call out in a forthcoming review.

I also fixed the broken `testAllAttachments()` test, which scans the classpath for attachments to validate (as opposed to the `testSpecificAttachments()` test, which uses a hard-coded list of attachments).  There are some issues with `testAllAttachments()` which make me question its usefulness.  See below for a more in-depth discussion.

I think it might be easier to review this PR commit-by-commit, as they are pretty focused, and the commit messages establish the context for the changes.

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -1,3 +1,3 @@
 checkstyleIntegTestMaxWarnings=0
-checkstyleMainMaxWarnings=3537
+checkstyleMainMaxWarnings=3539
```
Comment (ssoloff) [inline gradle.properties]: I introduced some new methods that had abbreviation violations (they contained `AA` in the name).  I was unable to change the name due to how attachment methods are invoked reflectively.  So I just bumped the warning count rather than try to introduce some unrelated warning fixes.

File: src/main/java/games/strategy/engine/data/annotations/GameProperty.java [hunk 2]
```diff
@@ -36,4 +36,17 @@
...
+   * with other non-virtual properties.
+   *
+   * <p>
+   * A virtual property is typically used for so-called "convenience" properties that are used to initialize multiple
+   * properties at once. Properties of this type are not required to have get, reset, or clear methods.
+   * </p>
+   *
+   * @return {@code true} if the property is virtual; otherwise {@code false}.
+   */
+  boolean virtual() default false;
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/annotations/GameProperty.java]: This was my creative way to work around validation errors for properties that can be set in the XML but do not have a true backing field because they are typically an alias to conveniently set one or more properties at once.  I'm not sure if calling them "virtual properties" is the best choice for a name.

File: src/main/java/games/strategy/triplea/attachments/UnitAttachment.java [hunk 3]
```diff
@@ -1367,7 +1385,7 @@ public void setBombard(final Integer s) {
     m_bombard = s;
   }
 
-  public int getBombard(final PlayerID player) {
+  public int getBombard() {
```
Comment (ssoloff) [inline src/main/java/games/strategy/triplea/attachments/UnitAttachment.java]: This parameter was unused.

File: src/main/java/games/strategy/triplea/attachments/CanalAttachment.java [hunk 4]
```diff
@@ -152,7 +152,7 @@ public void setExcludedUnits(final HashSet<UnitType> value) {
     m_excludedUnits = value;
   }
 
-  public HashSet<UnitType> getExcludedUnits(final GameData data) {
+  public HashSet<UnitType> getExcludedUnits() {
```
Comment (ssoloff) [inline src/main/java/games/strategy/triplea/attachments/CanalAttachment.java]: This parameter was unused.

File: src/test/java/games/strategy/engine/data/annotations/ValidateAttachmentsTest.java [hunk 5]
```diff
@@ -140,32 +130,32 @@ public void testSpecificAttachments() {
...
-    // find the classes folder
-    final URL url = getClass().getResource("/");
-    File file = null;
-    try {
-      file = new File(url.toURI());
-      file = new File(file.getParent(), "classes");
-    } catch (final URISyntaxException e) {
-      fail(e.getMessage());
-      ClientLogger.logQuietly(e);
+  public void testAllAttachments() throws Exception {
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/data/annotations/ValidateAttachmentsTest.java]: This test was previously broken because it was using the wrong classpath (probably prior to the switch to Gradle).  I fixed it so that it works when run from both Eclipse and Gradle (I don't know if it works under IntelliJ).  However, it has several issues:

* It loads nearly _every_ TripleA class on the classpath.
    * It can be configured to skip certain problematic classes via the `SKIPCLASSES` field, but that smells a bit and is yet another list of strings to be maintained.
    * Because it attempts to load some UI classes, it cannot be run in a headless environment.  Thus, it is skipped when this condition is true (i.e. it won't run on Travis, where it would probably be the most useful).
* It is a relatively long-running unit test because of the previous issue (it averages about 10 seconds on my machine).

I'm not sure if it's worth keeping this test given that we can provide a whitelist of classes to test in `testSpecificAttachments()`.  The obvious drawback is if someone adds a new attachment, and forgets to update the list in `getKnownAttachmentClasses()`, the new attachment could go untested.  This has apparently happened recently, as the `UserActionAttachment` class was not added to this list when that attachment was added to the codebase.

File: src/main/java/games/strategy/triplea/attachments/CanalAttachment.java (hunk 4 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/triplea/attachments/CanalAttachment.java]: Is the implementation type required or is Set sufficient?

Comment (ssoloff) [inline src/main/java/games/strategy/triplea/attachments/CanalAttachment.java]: Unfortunately, the tests verify that the property getter method return type exactly matches the associated field type. 😞  You can see this throughout all the attachment classes.  It appears to have been this way since the test was first introduced six years ago.

**Comment by ssoloff** [comment]
Rebased against `master` at b44d207 (1.9.0.0.5886).

**Comment by ssoloff** [comment]
Rebased against `master` at 0b40d60 (1.9.0.0.5925).

**Comment by Cernelius** [comment]
I understand almost nothing of what this is aimed at or what's the matter with those 500 errors, but 2012 sounds like you should be wanting to have a review from @veqryn, if possible.
You know that properties are supposed not to be validated to support custom ones for conditions right? Actually, not even sure if this is related to what you are doing.

**Comment by RoiEXLab** [comment]
@Cernelius You misunderstood something :)
By tests we mean code that is executed when submitting a PR to ensure nothing will break when applying the proposed changes.
In other words: This doesn't affect the triplea code which is delivered to the end user in any way ^^

**Comment by veqryn** [comment]
Looks good to me.  I think we disabled it because we didn't think of adding the 'virtual' flag you just created.  👍

**Comment by ssoloff** [comment]
Rebased against `master` at d6dc810 (1.9.0.0.5983).

**Comment by ssoloff** [comment]
Rebased against `master` at c690912 (1.9.0.0.6002).
